### DISCUSSION TITLE
Lesson 6: Brownie Fund Me - CompilerError: File outside of allowed directories

### DISCUSSION THREAD

**Discussion by ace-brown** [smartcontractkit/full-blockchain-solidity-course-py#213] [Q&A] [answered]
Hi guys.
At around hours 5:10 I encountered this error.
CompilerError: solc returned the following errors:

contracts/FundMe.sol:5:1: ParserError: Source "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol" not found: File outside of allowed directories.
import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";

![Capture](https://user-images.githubusercontent.com/79329055/137617425-3ccad4b6-e3d2-4bf2-b214-093a793d103a.JPG)

my brownie.config.yamal

![Capture](https://user-images.githubusercontent.com/79329055/137617765-027bf6f4-8f21-42e5-8427-f84d67fd1793.JPG)

**Comment by ace-brown** [comment]
**This is the FundMe.sol**

```javascript
// SPDX-License-Identifier: MIT

pragma solidity ^0.6.6;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@chainlink/contracts/src/v0.6/vendor/SafeMathChainlink.sol";

contract FundMe {
    using SafeMathChainlink for uint256;
    
    mapping(address => uint256) public addressToAmountFunded;
    address[] public funders;
    address public owner;
    
    constructor() public {
        owner = msg.sender;
    }
    
    function fund() public payable {
        uint256 minimumUSD = 50 * 10 ** 18;
        require(getConversionRate(msg.value) >= minimumUSD, "You need to spend more ETH!");
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
    
    function getVersion() public view returns (uint256){
        AggregatorV3Interface priceFeed = AggregatorV3Interface(0x8A753747A1Fa494EC906cE90E9f37563A8AF630e);
        return priceFeed.version();
    }
    
    function getPrice() public view returns(uint256){
        AggregatorV3Interface priceFeed = AggregatorV3Interface(0x8A753747A1Fa494EC906cE90E9f37563A8AF630e);
        (,int256 answer,,,) = priceFeed.latestRoundData();
         return uint256(answer * 10000000000);
    }
    
    // 1000000000
    function getConversionRate(uint256 ethAmount) public view returns (uint256){
        uint256 ethPrice = getPrice();
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        return ethAmountInUsd;
    }
    
    modifier onlyOwner {
        require(msg.sender == owner);
        _;
    }
    
    function withdraw() payable onlyOwner public {
        msg.sender.transfer(address(this).balance);
        
        for (uint256 funderIndex=0; funderIndex < funders.length; funderIndex++){
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);
    }
}
```

### **this the brownie-config.yaml:**


```
dependencies:
  - smartcontractkit/chainlink-brownie-contracts@1.1.1
compiler:
  solc:
    remappings:
      - "@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1"

dotenv: .env

wallets:
  from_key: ${PRIVATE_KEY}
```

### **And this is the error:**

```Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS E:\Web development\My projects\Blockchain projects\brownie_fund_me> brownie compile
INFO: Could not find files for the given pattern(s).
Brownie v1.17.0 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.6.12
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

contracts/FundMe.sol:5:1: ParserError: Source "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol" not found: File outside of allowed directories.
import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
^--------------------------------------------------------------------------^

contracts/FundMe.sol:6:1: ParserError: Source "@chainlink/contracts/src/v0.6/vendor/SafeMathChainlink.sol" not found: File outside of allowed directories.
import "@chainlink/contracts/src/v0.6/vendor/SafeMathChainlink.sol";
^------------------------------------------------------------------^

PS E:\Web development\My projects\Blockchain projects\brownie_fund_me>
```

**Reply by PatrickAlphaC** [reply]
really odd. Everything looks right. This error means it can't find that folder. 

what version of python are you using?

**Reply by ace-brown** [reply]
Yeah that's really odd. I've got installed python version 2.7, 3.9 and 3.10 on my 64 win10.  But on the system environment, I set all my paths to python 3.9.

**Comment by cromewar** [comment, marked as answer]
It's not `brownie_config.yaml` is `brownie-config.yaml` with `-` so that's why your brownie can't find the configuration, give it a try ;)
Best regards.

![image](https://user-images.githubusercontent.com/20199180/137805353-09455449-6878-4d92-87dc-addac8d48098.png)

**Reply by ace-brown** [reply]
That's it. It was the underscore keeping me form going on for 2 days. lol. BTW, thank you so much man.

**Reply by cromewar** [reply]
You are welcome, I'm glad it helped.

**Reply by pgraciae** [reply]
Had the same problem here! Thanks!

**Comment by s1n7** [comment]
I have the same issue. I spelled the "brownie-config.yaml" folder correctly. However, the error still occurs. 
My files have the same content as the ones posted here. The only difference is that I use a virtual environment, so my folder structure is a little bit different:
![image](https://user-images.githubusercontent.com/65351091/137935137-7b902ba8-3bbc-47bb-94d4-c1a272636f57.png)

The "dev" folder contains the contracts, scripts etc. and the "virtualenv" folder contains everything that belongs to the virtual environment. 
Can someone help?

**Reply by s1n7** [reply]
additional information: my solidity version seems to be the problem:
![image](https://user-images.githubusercontent.com/65351091/137935548-2609c7e4-db5f-4613-bcff-b81905e3d2af.png)

**Reply by cromewar** [reply]
just right, change the global compiler version - remote and set it the version you need. If you have compiler error also attach them here please.

**Reply by s1n7** [reply]
you mean right clicking on that and then select "Solidity: Change global compiler version (remote)"? And which version should I select? 
I'm sorry, I do not fully understand how to use the versions yet...

**Reply by cromewar** [reply]
There are some thing to take in consideration here, the extension provided by Juan Blanco is by far the best for Solidity development, however it has its faults, for example let's suppose you are going to work with a project using solidity version 0.8.0, so you should change the compiler version to it and it should work, but, what if you import another file which uses an anterior version? for example a Mock, and the mock has something like this:

`pragma solidity ^0.6.0;`

Event tough the compiler will work just fine, because you have the `^` symbol (meaning that you could use that version of the compiler or any another one higher) the VScode Extension is going to complain about the version you are using are not the same as in the file.

My recommendation, just use the compiler version according to your main contract file, ^0.6.6 or 0.8.0 in most cases (that I've seen) and for test contracts don't pay attention to vscode complaining to them (as long as your main compiler is higher to the required version).

Remember changing between version is as simple as right click and select the option.

![image](https://user-images.githubusercontent.com/20199180/138133965-055eb468-2f68-4914-bfb7-fdba4357f1ab.png)

**Comment by Xingfu-Yi** [comment]
Perhaps it's a grammar mistake in brownie-config.yaml , i wrongly type the word commpiler and that's the case

**Reply by cromewar** [reply]
Nice hint @steven2077

**Comment by PP-AR-T** [comment]
I got the same problem, and can't find the reason no one of the listed above

**Comment by tredw** [comment]
Hi - I have the same issue and the solutions above have not worked.

![Fundme](https://user-images.githubusercontent.com/97856144/152890204-d0eb5b44-866f-4d5c-88f6-6b3357a0b3cb.PNG).

Below is the brownie-config.yaml file:

![yaml](https://user-images.githubusercontent.com/97856144/152890634-a141f08e-6d3a-4997-abaa-d7ba8cec2df9.PNG)

I also added this to settings.json file
 "solidity.remappings": [
        "@chainlink/=/Users/Tim/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@1.1.1",
        "@openzeppelin/=/Users/Tim/.brownie/packages/OpenZeppelin/openzeppelin-contracts@4.4.0"

**Reply by cromewar** [reply]
@tredw it seems your `brownie-config.yaml` is inside your test folder, is it?
if that's true, it must be on the project root directory.

**Reply by tredw** [reply]
Thanks!  You're right much appreciated.  Lesson I've learnt from this is sometimes when stuck just need to re-watch video from the beginning of lesson you're on and go through the coding process again to hopefully catch what you missed the first time.

**Reply by Yadu-blockchain** [reply]
Thank you so much man! I was working with this problem for a very long time. Finally.

**Comment by realglass** [comment]
Hi! I am a complete noob at this but learning quickly. It seems I got suck here and can't figure out what is the problem. Any help pls??

![image](https://user-images.githubusercontent.com/101928609/159118422-5ddeb435-51ed-4765-b0c5-2b3aae141f79.png)

here is my brownie-config.yaml

![image](https://user-images.githubusercontent.com/101928609/159118451-42e7177e-18de-4c8c-a877-5bb449a49525.png)

**Reply by cromewar** [reply]
Hello @andrewvella1991 I don't see the mockContract on your contracts folder, you must create a folder called tests and put there the MockContract.

**Reply by tolivern** [reply]
Maybe it is because you wrote "dependecies" instead of "dependencies"...? (I am noob as well)

**Reply by cromewar** [reply]
yes you are right @tolivern 100% missed it

**Reply by Madhur-Prakash** [reply]
thnx @tolivern, you are indeed correct

**Comment by Abdel364** [comment]
Hello Please can some help me with this issue, i got the some issue of you guys but i don't know where is the problem 

This is the contract:
<img width="1023" alt="Screenshot 2022-03-24 at 19 43 34" src="https://user-images.githubusercontent.com/100959384/159989066-eb96b2de-3822-4d45-bd7f-b417419abc3f.png">
<img width="1436" alt="Screenshot 2022-03-24 at 23 25 08" src="https://user-images.githubusercontent.com/100959384/160020465-50ed958b-5171-47d0-8e04-784c7c625620.png">
<img width="1439" alt="Screenshot 2022-03-24 at 19 47 11" src="https://user-images.githubusercontent.com/100959384/159989114-9f88d7d4-b3fa-4af1-ace5-440158ebd883.png">
<img width="1438" alt="Screenshot 2022-03-24 at 19 47 23" src="https://user-images.githubusercontent.com/100959384/159989123-1baab951-ab4d-487d-b6ae-db1a8db791c9.png">
<img width="1063" alt="Screenshot 2022-03-24 at 20 03 01" src="https://user-images.githubusercontent.com/100959384/159991345-9732d871-aa14-41b7-837d-d2d117515a74.png">

**Reply by rohit9888** [reply]
it is remappings not remapping. add 's'

**Reply by harshpreetwmk** [reply]
thanks brother love you

**Comment by ajimoti** [comment]
![Screenshot 2022-04-30 at 4 18 02 PM](https://user-images.githubusercontent.com/26599467/166111541-b461af1f-4683-4fb4-a314-f69fbdab1b9b.png)

I had the same error, apparently, I didn't leave a space between the `-` and `'@chainlink=smartcontractkit/chainlink-brownie-contracts@1.1.1'` in my remappings.

Solved it by adding the space
![Screenshot 2022-04-30 at 4 18 58 PM](https://user-images.githubusercontent.com/26599467/166111578-8f52eee9-49f8-40e4-93a2-0d01789912dd.png)
